### PR TITLE
Added storage_profile support

### PR DESCRIPTION
**Pull request by dn1s**
- Added vm description option.
- Changed api call which creates vapp + vm from instantiate_template to
 compose vapp.
- Removed network_href variable which is not of use anymore since the
 api call changed.
- Changed template_name attribute from being an Optional parameter to
 Required parameter.

### CODE REVIEW

**Comment by bobbydeveaux** [comment]
Just running the acceptance tests in CI.

**Comment by bobbydeveaux** [comment]
Addresses point 5 of https://github.com/terraform-providers/terraform-provider-vcd/issues/9

File: vcd/resource_vcd_vapp.go [hunk 1]
```diff
@@ -92,7 +96,7 @@ func resourceVcdVAppCreate(d *schema.ResourceData, meta interface{}) error {
 
 		catalogitem, err := catalog.FindCatalogItem(d.Get("template_name").(string))
 		if err != nil {
-			return fmt.Errorf("Error finding catalog item: %#v", err)
+			return fmt.Errorf("Error finding catelog item: %#v", err)
```
Comment (robcoward) [inline vcd/resource_vcd_vapp.go]: This is a typo surely ?

File: vcd/resource_vcd_vapp.go [hunk 2]
```diff
@@ -17,61 +17,65 @@ func resourceVcdVApp() *schema.Resource {
...
+				Required: true,
 				ForceNew: true,
 			},
 
-			"catalog_name": &schema.Schema{
-				Type:     schema.TypeString,
-				Optional: true,
-			},
-
-			"network_href": &schema.Schema{
```
Comment (robcoward) [inline vcd/resource_vcd_vapp.go]: What is the reasoning for removing the network_href schema attribute ?

**Comment by bobbydeveaux** [comment]
1 of the tests have failed due to catalog_name now being required. In the master branch there is new functionality that allows a raw vApp to be created. You'll need to ensure these tests still pass before we can merge.

File: vcd/resource_vcd_vapp.go (hunk 1 above)
Comment (dn1s) [inline vcd/resource_vcd_vapp.go]: True

**Comment by dn1s** [comment]
I'll fix this in the next days.

File: vcd/resource_vcd_vapp.go (hunk 2 above)
Comment (dn1s) [inline vcd/resource_vcd_vapp.go]: I just copy + pasted my changes I did last time against terraform repo. I see that it's now used in a different way. Have to check again if the error I had before still exists. Thanks for pointing this out!

**Comment by bobbydeveaux** [comment]
Awesome, thanks for the PR btw - great work :) Ping me when the changes are done and I'll run the full suite again :)

**Comment by dn1s** [comment]
Updated my pull request

**Comment by bobbydeveaux** [comment]
@dn1s where is your code checked out to? terraform-providers/terraform-provider-vcd ?

**Comment by bobbydeveaux** [comment]
CI is just running these tests on a fresh VDC.

**Comment by dn1s** [comment]
It ran 2 ci tests without the .travis.yaml but after I rebased on the latest changes of master it didn't run any ci test anymore. But never mind that. Is it working?

**Comment by bobbydeveaux** [comment]
Afraid not:

------- Stdout: -------
=== RUN   TestAccVcdVAppVm_Basic
--- FAIL: TestAccVcdVAppVm_Basic (115.49s)
    testing.go:428: Step 0 error: Error applying: 1 error(s) occurred:
        
        * vcd_vapp.foobar: 1 error(s) occurred:
        
        * vcd_vapp.foobar: Error finding VApp: &errors.errorString{s:"can't find vApp: foobar"}
FAIL

------- Stdout: -------
=== RUN   TestAccVcdVApp_PowerOff
--- FAIL: TestAccVcdVApp_PowerOff (112.23s)
    testing.go:428: Step 0 error: Error applying: 1 error(s) occurred:
        
        * vcd_vapp.foobar: 1 error(s) occurred:
        
        * vcd_vapp.foobar: Error finding VApp: &errors.errorString{s:"can't find vApp: foobar"}
FAIL

**Comment by bobbydeveaux** [comment]
I'm just re-running CI on master just to confirm master is all okay.

**Comment by bobbydeveaux** [comment]
Yeah, sorry @dn1s - master is passing, but I've ran this PR 3 times (it can be flaky..) but it keeps failing with the above errors. Master passed first time.

**Comment by dn1s** [comment]
Found the issue but I don't get it yet have to test it more.

**Comment by dn1s** [comment]
Out of time shortage I won't implement this feature into the new vm resource. But feel free to reuse my code for vapp.

**Comment by dn1s** [comment]
Could you test this again, please?
There is the possibility that this fails because the test can't find a suitable storage profile if this is the case
just provide me with a default storage profile and I'll update my pull request.

**Comment by bobbydeveaux** [comment]
Sorry @dn1s - just merged the other PR as tests have just passed on that one and now there are some conflicts :(

**Comment by dn1s** [comment]
Rebased on latest changes

**Comment by dn1s** [comment]
Bump @bobbydeveaux

**Comment by bobbydeveaux** [comment]
Sorry @dn1s - will re- run test suite :)

**Comment by bobbydeveaux** [comment]
Not managed to get it to pass yet, but the failures are pointing at an issue with the environment. Will try on a fresh setup!

**Comment by dn1s** [comment]
Any update here?

**Comment by bobbydeveaux** [comment]
Sorry @dn1s - didn't have any luck refreshing the environment. Just managed that now and running the test suite again. Will let you know the result! :)

**Comment by bobbydeveaux** [comment]
Sorry @dn1s - tests still failing, but struggling to get master to pass either right now (master was last successful 12th July and there have been no changes. if master isn't passing, there's no hope for this branch. 

I have a couple of hours today to spend looking at this. Will report back before the day is over!

**Comment by dn1s** [comment]
@bobbydeveaux you failed with reporting. :P

**Comment by bobbydeveaux** [comment]
D'oh. Sorry. Issue with he edge gateway in vCloud. Simone at UKCloud is looking into it. Will chase this AM :)

**Comment by bobbydeveaux** [comment]
@dn1s 

Master branch success! 

OK	 TestAccVcdVAppVm_Basic 	7m:13s,549ms 	9
OK	 TestAccVcdVApp_PowerOff 	4m:55s,594ms 	8
OK	 TestAccVcdVAppRaw_Basic 	4m:10s,705ms 	7
OK	 TestAccVcdNetwork_Basic 	1m:25s,404ms 	5
OK	TestAccVcdFirewallRules_basic 	1m:03s,467ms 	4
OK	TestAccVcdVpn_Basic 	56s,229ms 	3
OK	 TestAccVcdSNAT_Basic 	56s,181ms 	6
OK	TestAccVcdDNAT_Basic 	55s,512ms 	1
OK	TestAccVcdDNAT_tlate 	55s,446ms 	2

Will now re-reun this PR!

**Comment by bobbydeveaux** [comment]
Finally, some good news @dn1s 

OK	 TestAccVcdVAppVm_Basic 	7m:15s,499ms 	9
OK	 TestAccVcdVApp_PowerOff 	5m:34s,860ms 	8
OK	 TestAccVcdVAppRaw_Basic 	3m:48s,661ms 	7
OK	 TestAccVcdNetwork_Basic 	1m:23s,909ms 	5
OK	TestAccVcdFirewallRules_basic 	1m:03s,046ms 	4
OK	TestAccVcdVpn_Basic 	56s,892ms 	3
OK	TestAccVcdDNAT_tlate 	55s,493ms 	2
OK	TestAccVcdDNAT_Basic 	55s,230ms 	1
OK	 TestAccVcdSNAT_Basic 	51s,378ms 	6

**Comment by bobbydeveaux** [comment]
Just had a bug report following this @dn1s 

```FYI - When I set `power_on     = false` in `vcd_vapp` resource it still powers on```

I can see the bug, I shall fix now :)

**Pull request merged**